### DISCUSSION TITLE
Open browser after spinning up containers?

### DISCUSSION THREAD

**Discussion by adameichelkraut** [batect/batect#1222] [Ideas]
Hey! I've been using Batect now for a bit, but I've been wondering about this one: is there a way to have batect know that we are spinning up a web interface that we would want to open a browser to? For example, usually when you do a `create-react-app` and then do `npm start`, NPM would open the browser for you once the development server has started.

However, I started to wonder if it was appropriate for Batect to handle. To me, it seems like a nice efficiency to have in my `batect.yml`, but I could understand it being interesting to justify a "who should be the one to do that".

Personally, I've been trying to get to a cool spot with using batect to manage architecture diagrams using drawio's docker images (https://hub.docker.com/r/jgraph/drawio/). It would be awesome if I could have my architecture diagram automatically spin up with a `./batect diagrams` and then it automatically open the browser to the location I'd like it to open.

Interested to hear folks' thoughts!

**Comment by adameichelkraut** [comment]
also, if it makes sense I could give this one a crack at contributing :)

**Comment by charleskorn** [comment]
It's an interesting idea, and I think could be implemented through something like what #78 proposes - what do you think?

**Comment by adameichelkraut** [comment]
That sounds great! Here's a thought:

what if we specified tasks like this:

```
tasks:
    diagrams:
      post-requisites:
          - openBrowser
      run:
          container: drawio
    
    openBrowser:
        run:
            container: host
            # or
            executeOnHost: true
            command: xdg-open https://localhost:3000/
```

**Reply by adameichelkraut** [reply]
Actually wait, no because the containers above would be long lived, hypothetically so in my instance I would only need the host task to run in the prerequisites 

```
tasks:
    diagrams:
      prerequisites:
          - openBrowser
      run:
          container: drawio
    
    openBrowser:
        run:
            container: host
            # or
            executeOnHost: true
            command: xdg-open https://localhost:3000/
```

**Comment by adameichelkraut** [comment]
thinking about this a little more though... the biggest reason I use batect is to make a platform independent development environment. If we started to add "host tasks" in here, I feel like we might be running the risk of turning Batect into something it's specifically not: a generalized task executor. 

In the above example, the 'xdg-open' would most likely work for most Unix/Linux platforms - but excludes probably Windows-based hosts. In this case, we would want to define some form of "If this host, do this command, if Windows, do this other thing"... which could get pretty hairy.

But maybe having predefined tasks such as "open browser", or something would be nice? And also how to implement your own custom "host task" that is platform independent?

**Reply by charleskorn** [reply]
> thinking about this a little more though... the biggest reason I use batect is to make a platform independent development environment. If we started to add "host tasks" in here, I feel like we might be running the risk of turning Batect into something it's specifically not: a generalized task executor.
> 
> In the above example, the 'xdg-open' would most likely work for most Unix/Linux platforms - but excludes probably Windows-based hosts. In this case, we would want to define some form of "If this host, do this command, if Windows, do this other thing"... which could get pretty hairy.
> 

Yeah that's exactly where it gets quite tricky. And we can't even rely on things like pushing this complexity out to a Bash script that you write, because there's no guarantee that Windows hosts have Bash installed.

> But maybe having predefined tasks such as "open browser", or something would be nice?

This seems like a very specific situation (albeit a very common one), so I'm not sure whether it makes sense to open a potential can of worms for Batect in supporting stuff like this. On the one hand, I can definitely see the value in this and the time it would save, but on the other hand, as you point out, there's a whole bunch of platform-specific differences that have to be taken into consideration.

**Reply by charleskorn** [reply]
tl;dr: I'm interested to see what others think before making a call either way.

**Reply by XI1998-swapnilmarathe** [reply]
Here can we just have a simple functionality that allows user to run commands post containers are healthly like what @adameichelkraut said.

If that command eg: sh, notepad, or any script is present in host machine or not should not be concern for batect.

User is responsible to have things installed on their host machine. They should correctly define commands like we have delegated responsibility for commands in docker internal containers. 

eg:
`./batect mytask`

```
mytask:
  run:
    command: Idontcarewhatcommand
```

it will just call internally like `exec Idontcarewhatcommand`
